src/java.base/share/classes/javax/crypto/KEMSpi.java

```diff
@@ -0,0 +1,244 @@
+         *          the specified algorithm
+         * @throws DecapsulateException if an error occurs during the
+         *          decapsulation process
+         * @throws ArrayIndexOutOfBoundsException if {@code from < 0},
+         *          {@code from > to}, or {@code to > secretSize()}
+         * @throws NullPointerException if {@code encapsulation} or
+         *          {@code algorithm} is {@code null}
+         * @throws UnsupportedOperationException if the combination of
+         *          {@code from}, {@code to}, and {@code algorithm}
+         *          is not supported by the decapsulator
+         * @see KEM.Decapsulator#decapsulate(byte[], int, int, String)
+         */
+        SecretKey engineDecapsulate(byte[] encapsulation, int from, int to, String algorithm)
+                throws DecapsulateException;
+
+        /**
+         * Returns the size of the shared secret.
+         *
+         * @return the size of the shared secret as a finite non-negative integer
+         * @see KEM.Decapsulator#secretSize()
+         */
+        int engineSecretSize();
+
+        /**
+         * Returns the size of the key encapsulation message.
+         *
+         * @return the size of the key encapsulation message as a finite non-negative integer
+         * @see KEM.Decapsulator#encapsulationSize()
+         */
+        int engineEncapsulationSize();
+    }
+
+    /**
+     * Creates a KEM encapsulator on the KEM sender side.
+     *
+     * @param pk the receiver's public key, must not be {@code null}
+     * @param spec the optional parameter, can be {@code null}
+     * @param secureRandom the source of randomness for encapsulation.
+     *                     If {@code null}, the implementation must provide
+     *                     a default one.
+     * @return the encapsulator for this key
+     * @throws InvalidAlgorithmParameterException if {@code spec} is invalid
+     *          or one is required but {@code spec} is {@code null}
+     * @throws InvalidKeyException if {@code pk} is invalid
+     * @see KEM#newEncapsulator(PublicKey, AlgorithmParameterSpec, SecureRandom)
+     */
+    EncapsulatorSpi engineNewEncapsulator(PublicKey pk,
+            AlgorithmParameterSpec spec, SecureRandom secureRandom)
+            throws InvalidAlgorithmParameterException, InvalidKeyException;
+
+    /**
+     * Creates a KEM decapsulator on the KEM receiver side.
+     *
+     * @param sk the receiver's private key, must not be {@code null}
+     * @param spec the parameter, can be {@code null}
+     * @return the decapsulator for this key
+     * @throws InvalidAlgorithmParameterException if {@code spec} is invalid
+     *          or one is required but {@code spec} is {@code null}
+     * @throws InvalidKeyException if {@code sk} is invalid
+     * @see KEM#newDecapsulator(PrivateKey, AlgorithmParameterSpec)
+     */
+    DecapsulatorSpi engineNewDecapsulator(PrivateKey sk, AlgorithmParameterSpec spec)
```

Comment:
Bit of a nit, but suggest changing variable names to be more descriptive: `privKey` (or `privateKey`) and `pubKey` (or `publicKey`), for the `engineEncapsulate` method. I assume `sk` is taken from RFC 9180, but I don't think we need to use the same variable names.